src/lib.rs

```diff
@@ -1,22 +1,37 @@
-use data_structures::{Location, Region};
-use geo::{coord, point, Contains, LineString, Polygon};
+use data_structures::{Coordinates, Location, MatchedResult, Region};
+use geo::{coord, point, Contains};
 
 pub mod data_structures;
 
-pub fn match_locations_to_regions(locations: &Vec<Location>, regions: &mut Vec<Region>) {
-    for region in regions {
-        for polygon in &region.polygons {
-            let line_string: LineString<f64> = LineString::new(polygon.vertices.iter()
-                .map(|vert| coord! { x: vert.longitude.val(), y: vert.latitude.val() })
-                .collect()
-            );
-            let geo_polygon = Polygon::new(line_string, vec![]);
-            region.matched_locations.extend(locations.iter()
-                .filter(|location| geo_polygon.contains(&point!(x: location.coordinates.longitude.val(), y: location.coordinates.latitude.val())))
+fn create_linestring_from_coord_vec(coords: &Vec<Coordinates>) -> geo::LineString {
+    geo::LineString::new(coords.iter().map(|vert| coord! { x: vert.longitude.val(), y: vert.latitude.val() }).collect())
+}
+
+fn create_geopoint_from_loc(location: &Location) -> geo::Point {
+    point!(x: location.coordinates.longitude.val(), y: location.coordinates.latitude.val())
+}
+
+pub fn match_locations_to_regions(locations: &Vec<Location>, regions: &Vec<Region>) -> Vec<MatchedResult> {
+    regions.iter().map(
+        |region|
+        MatchedResult { 
+            region: region.name.clone(),
+            matched_locations: region.polygons.iter()
+                .map(
```

Comment:
I think you don't need two `map`s, you could create a single `map` operation, create a polygon , and then use it for searching.

Reply:
Yes, I checked it right now, it works